src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5179,6 +5184,12 @@ void learnFromPostConditions(BoundExpression argument, FlowAnalysisAnnotations p
                         LearnFromNullTest(argument, ref StateWhenFalse);
                     }
                 }
+
+                bool setNotNullFromParameterType()
+                    => !argument.IsSuppressed
+                    && ((!parameterType.Type.IsTypeParameterDisallowingAnnotation()
+                            && parameterType.NullableAnnotation.IsNotAnnotated())
+                        || parameterOriginalDefinitionType.Type is TypeParameterSymbol { HasNotNullConstraint: true });
```

Comment:
>parameterOriginalDefinitionType [](start = 27, length = 31)

I didn't understand why we need to use the original definition here.
If we're dealing with an unsubstituted method, then the parameter is `TNotNull` which we can check here (1).
If we're dealing with a substituted method, then the parameter was substituted with an unannotated type (2) or a safety diagnostic was reported on the substitution (3). Is scenario (3) what prompted you to use the original definition here?

Reply:
In scenario (3) adding this check allows us to update argument state to not-null which reduces cascading diagnostics in a similar fashion to updating argument state when the parameter type disallows null.

```cs
class C
{
    void M<T>(T t) where T : notnull { }

    void M1(string? s)
    {
        M(s); // warning: 'string?' does not meet the 'notnull' constraint
        s.ToString(); // in master we get another warning here, but this is avoidable.
    }
}
```

It occurs to me that if we do this here, we should probably do something similar for the 'class' constraint.

Reply:
I'd be okay if this we didn't handle this cascading diagnostic at all, or if we split that to a different PR. We would report a warning on the constraint, then one warning on passing a bad value to `M(s)` but no warning on `s.ToString()`.

Reply:
Why avoid reporting a warning for `s.ToString();` in the example? We don't know how the argument to `M<T>(T t)` is used. And we don't skip warnings in non-generic scenarios:
```C#
class C
{
    void M(string s) { }

    void M1(string? s)
    {
        M(s); // warning
        s.ToString(); // warning
    }
}
```

Reply:
@RikkiGibson From discussion with @cston, it feels like the second part of the PR should be reverted and not even addressed in a follow-up PR, as it is addressing a different scenario than the original point of this PR and what was discussed in LDM.
The main rationale from LDM is that we should assume that methods with non-nullable parameters really cannot handle null (one doesn't need to add an attribute) to express that. There was no discussion of avoiding cascades from violated constraints.

Reply:
I agree, it seems like this part of the change needs to be considered separately. I think I just misinterpreted the suggestion to add a `T : notnull` test.

Reply:
> The main rationale from LDM is that we should assume that methods with non-nullable parameters really cannot handle null

It does feel like this is also the case for type parameters constrained to `notnull` or `class`. However, whether to change anything in nullable analysis for this scenario is up to the LDM.